src/addrdb.cpp

```diff
@@ -199,11 +199,13 @@ bool CBanDB::Write(const banmap_t& banSet)
 
 bool CBanDB::Read(banmap_t& banSet, bool& dirty)
```

Comment:
The `dirty` argument is used to not write the json file again if it was read successfully and nothing needed to be swept. Happy to drop it, as it simplifies and writing the json once per startup shouldn't cost too much.

Reply:
Yes, since `Read()` does not do any sweeping its return value is enough to tell whether the read was successful or not. Or, in other words - with this PR it is not possible `Read()` to return `true` and set its second argument (`dirty`) to `true`.

Something like this should work (not tested):
```diff
diff --git i/src/addrdb.cpp w/src/addrdb.cpp
index fbffd5c959..c3e224ee83 100644
--- i/src/addrdb.cpp
+++ w/src/addrdb.cpp
@@ -194,25 +194,22 @@ bool CBanDB::Write(const banmap_t& banSet)
     for (const auto& err : errors) {
         error("%s", err);
     }
     return false;
 }
 
-bool CBanDB::Read(banmap_t& banSet, bool& dirty)
+bool CBanDB::Read(banmap_t& banSet)
 {
     if (fs::exists(m_banlist_dat)) {
         LogPrintf("banlist.dat ignored because it can only be read by " PACKAGE_NAME " version 22.x. Remove %s to silence this warning.\n", m_banlist_dat);
     }
     // If the JSON banlist does not exist, then recreate it
     if (!fs::exists(m_banlist_json)) {
-        dirty = true;
         return false;
     }
 
-    dirty = false;
-
     std::map<std::string, util::SettingsValue> settings;
     std::vector<std::string> errors;
 
     if (!util::ReadSettings(m_banlist_json, settings, errors)) {
         for (const auto& err : errors) {
             LogPrintf("Cannot load banlist %s: %s\n", m_banlist_json.string(), err);
diff --git i/src/addrdb.h w/src/addrdb.h
index 0213a14af5..1e0ccb1f60 100644
--- i/src/addrdb.h
+++ w/src/addrdb.h
@@ -92,17 +92,15 @@ public:
     bool Write(const banmap_t& banSet);
 
     /**
      * Read the banlist from disk.
      * @param[out] banSet The loaded list. Set if `true` is returned, otherwise it is left
      * in an undefined state.
-     * @param[out] dirty Indicates whether the loaded list needs flushing to disk. Set if
-     * `true` is returned, otherwise it is left in an undefined state.
      * @return true on success
      */
-    bool Read(banmap_t& banSet, bool& dirty);
+    bool Read(banmap_t& banSet);
 };
 
 /**
  * Dump the anchor IP address database (anchors.dat)
  *
  * Anchors are last known outgoing block-relay-only peers that are
diff --git i/src/banman.cpp w/src/banman.cpp
index d2437e6733..c64a48a05a 100644
--- i/src/banman.cpp
+++ w/src/banman.cpp
@@ -15,13 +15,13 @@
 BanMan::BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time)
     : m_client_interface(client_interface), m_ban_db(std::move(ban_file)), m_default_ban_time(default_ban_time)
 {
     if (m_client_interface) m_client_interface->InitMessage(_("Loading banlist…").translated);
 
     int64_t n_start = GetTimeMillis();
-    if (m_ban_db.Read(m_banned, m_is_dirty)) {
+    if (m_ban_db.Read(m_banned)) {
         SweepBanned(); // sweep out unused entries
 
         LogPrint(BCLog::NET, "Loaded %d banned node addresses/subnets  %dms\n", m_banned.size(),
                  GetTimeMillis() - n_start);
     } else {
         LogPrintf("Recreating the banlist database\n");
```

Reply:
Thanks, will pick the patch. It is almost correct 🥲:


```
==4031966== Conditional jump or move depends on uninitialised value(s)
==4031966==    at 0x46011B: BanMan::DumpBanlist() (banman.cpp:44)
==4031966==    by 0x45E962: BanMan::BanMan(boost::filesystem::path, CClientUIInterface*, long) (banman.cpp:32)
==4031966==    by 0x15C119: make_unique<BanMan, boost::filesystem::path, CClientUIInterface *, long> (unique_ptr.h:962)
==4031966==    by 0x15C119: AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*) (init.cpp:1164)
==4031966==    by 0x141388: AppInit (bitcoind.cpp:226)
==4031966==    by 0x141388: main (bitcoind.cpp:269)
==4031966==

Reply:
Oh, why did you had to run valgrind on this? It was perfectly safe before that...